Add controller-id proposal (#5919)
prometheus-operator/prometheus-operator: Documentation/proposals/202309-controller-id.md

# Controller-ID Proposal

## Owners:
* [danielmellado](https://github.com/danielmellado)

## Related Tickets:
* [#4281](https://github.com/prometheus-operator/prometheus-operator/issues/4281)
* [#4498](https://github.com/prometheus-operator/prometheus-operator/pull/4498)

## Other Docs:
* n/a

### Introduction

This proposal aims to implement a solution to support multiple cluster-level

> Prometheus instances running concurrently without conflicting over the same CRD.

Reviewer: (nit) CRD -> custom resources. CRD is a definition of a custom resource (e.g. the schema definition)

This issue can significantly impact use cases where multiple Prometheus operator
instances run in parallel.

## Why

Currently, we encounter issues when users or different operators deploy Prometheus
alongside Prometheus Operator's instance. Even though User-Managed Workloads
(UWM) exists, some operators or users may deploy Prometheus directly from
manifests, Helm charts, or other rogue operators, leading to the creation of

> 'rogue' Prometheus instances alongside the operator's instance.

Reviewer: I don't think that Prometheus instances deployed without the operator are an issue. I would simplify the paragraph saying that the core issue is when different users deploy different instances of the Prometheus operator which try to reconcile the same resources.

of the CRD, causing disruptions to all pods.

The remediation for this scenario, where users deploy their Prometheus operator
instances in parallel, involves using one of the many CLI arguments, such as
`--deny-namespaces`, `--namespaces`, `--prometheus-instance-selector`, or

> `prometheus-instance-namespaces`.

Reviewer: ```suggestion
`prometheus-instance-namespaces` But this requires cooperation between the different parties and there's no way to ensure that a specific monitoring resource is managed only by a specific operator instance.
```

instances in parallel, involves using one of the many CLI arguments, such as
`--deny-namespaces`, `--namespaces`, `--prometheus-instance-selector`, or
`prometheus-instance-namespaces`.

> Although the initial discussion for this proposal considered adding an owner

Reviewer: should be listed in an alternatives section.

[zalando/postgres-operator](https://github.com/zalando/postgres-operator). When
an operator is configured with a specific "controller ID" value, it will only
reconcile resources that have a matching "controller ID" annotation.

Conversely, if the operator is not configured with a "controller ID," it will
skip all resources that have a "controller ID" annotation. More details can be
found in the
[zalando/postgres-operator documentation](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#operators-with-defined-ownership-of-certain-postgres-clusters).

# Goals

* Guarantee that a custom resource will be managed by a specific Prometheus
  operator instance.

> 

Reviewer: For non-goals, I'd add
*  it's not going to be an zero-configuration solution (e.g. it still requires work from the user deploying the operator + resource).
* this will be an opt-in feature

Reviewer 2: You mean "This will not be an opt-in feature"?

Author: I assume that Simon means that users would be able to not trigger the feature? Not sure why they wouldn't like to but just in case they would opt-out the config?

Reviewer: I mean to say that when you start the operator without any specific argument, it will attempt to reconcile all resources in all namespaces. But I realize that it's similar to the first non-goal.

Author: I'll remove the 2nd and put a clarification on the first one.

found in the
[zalando/postgres-operator documentation](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#operators-with-defined-ownership-of-certain-postgres-clusters).

## Goals

> * Add a similar "controller-id" based approach to PO.

Reviewer: I'd rather phrase it as: "Guarantee that a custom resource will be managed by a specific Prometheus operator instance"

* Users seeking to mitigate the impact of rogue Prometheus instances.

# Alternatives

Although the initial discussion for this proposal considered adding an owner
reference within the scope, the
[ControllerRef](https://github.com/kubernetes/design-proposals-archive/blob/acc25e14ca83dfda4f66d8cb1f1b491f26e78ffe/api-machinery/controller-ref.md)

> model does not directly address this problem.

Reviewer: ```suggestion
model does not directly address this problem because the `ControllerRef` model solves the problem of controllers that fight over controlled objects due to overlapping selectors (e.g. a ReplicaSet fighting with a ReplicationController over Pods because both controllers have label selectors that match those Pods)
```

# Controller-ID Proposal

* Owners:
  * [danielmellado](https://github.com/danielmellado)
* Related Tickets:
  * [#4281](https://github.com/prometheus-operator/prometheus-operator/issues/4281)
  * [#4498](https://github.com/prometheus-operator/prometheus-operator/pull/4498)
* Other Docs:
  * n/a

# Introduction

This proposal aims to implement a solution to support multiple cluster-level
Prometheus instances running concurrently without conflicting over the same

> custom resources.

Reviewer: This is probably obvious for all of us but the solution isn't going to be limited to the Prometheus resources and will also be available for Alertmanager and ThanosRuler (and any pod-based resource we may add in the future).

Author: I'll add a note about that, thanks!

custom resources. This solution isn't limited to the Prometheus resources as
it'll also be available for `AlertManager`and `ThanosRuler` ones, as well as for
any pod-based resource that could be added in the future.

This issue can significantly impact use cases where multiple Prometheus operator
instances run at the same time in the Kubernetes cluster.

# Why

Currently, we encounter issues when different users deploy different instances
of the Prometheus operator, that will try to reconcile the same resources.

In the worst-case scenario, these operators may not only compete for ownership
of the CRD resources but also attempt to rewrite or redeploy different versions
of the CRD, causing disruptions to all pods.

The remediation for this scenario, where users deploy their Prometheus operator
instances in parallel, involves using one of the many CLI arguments such as
`--deny-namespaces`, `--namespaces`, `--prometheus-instance-selector`  or
`prometheus-instance-namespaces`. But this requires cooperation between the
different parties and there's no way to ensure that a specific monitoring
resource is managed only by a specific operator instance.

# How

> After some research, we have identified a potential solution already

Reviewer: I forgot that we should give credit to @machine424 for the suggestion :)
```suggestion
After some research from @machine424, we have identified a potential solution already
```

Author: Totally agree, adding him back there, wasn't sure whether it should be on the proposal.

implemented by the
[zalando/postgres-operator](https://github.com/zalando/postgres-operator). When
an operator is configured with a specific "controller ID" value, it will only
reconcile resources that have a matching "controller ID" annotation.

Conversely, if the operator is not configured with a "controller ID," it will
skip all resources that have a "controller ID" annotation. More details can be
found in the
[zalando/postgres-operator documentation](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#operators-with-defined-ownership-of-certain-postgres-clusters).

# Goals

* Guarantee that a custom resource will be managed by a specific Prometheus
  operator instance.

## Audience

This proposal is relevant to the following audience:

* Users who provide Prometheus as a service and want to run multiple Prometheus
  operator instances in different namespaces.
* Users seeking to mitigate the impact of rogue Prometheus instances.

# Non-Goals

> * It's not going to be a zero-configuration solution, as it'll require work from

Reviewer: (nit)
```suggestion
* Provide a solution that works with user intervention. It'll require additional configuration from
```
